### PR TITLE
gpl: fix out of bounds vector access

### PR DESCRIPTION
**Pull request by gudeh**
This fix aims to prevent out-of-bounds access in a vector that stores tiles within GPL code.

The issue arises when the code iterates over instances and accesses tiles using the instances' coordinates. Should an instance be positioned too close to the border, the existing code would calculate an index exceeding the maximum permissible value. This seems to occur due to the size extension applied to border tiles.

In my testing, this problem manifested in a singular scenario: running ihp/ibex with Yosys version 0.36. The implementation of this fix has successfully addressed the issue.

FYI: within GPL code, GCells refer to instances, such as global cells, and do not denote grid cells. This is based on my current understanding.

### CODE REVIEW

**Comment by gudeh** [comment]
@precisionmoon

**Comment by rovinski** [comment]
Hi @gudeh,

Considering this is a rare edge case and you've found the root cause, would you mind creating a minimal test case to go along with the fix?

**Comment by precisionmoon** [comment]
Let me try my test with your changes.  Yes, as Austin suggested, adding a unit test will be very nice.

File: src/gpl/src/routeBase.cpp [hunk 1]
```diff
@@ -557,10 +557,16 @@ std::pair<bool, bool> RouteBase::routability()
...
+    int idxX = std::min((gCell->dCx() - tg_->lx()) / tg_->tileSizeX(),
+                        tg_->tileCntX() - 1);
+    int idxY = std::min((gCell->dCy() - tg_->ly()) / tg_->tileSizeY(),
+                        tg_->tileCntY() - 1);
 
-    Tile* tile = tg_->tiles()[idxY * tg_->tileCntX() + idxX];
+    size_t index = idxY * tg_->tileCntX() + idxX;
+    if (index >= tg_->tiles().size()) {
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (gudeh) [reply src/gpl/src/routeBase.cpp:568]: Yes. The continue is just an extra precaution.

**Comment by gudeh** [comment]
Hi everyone. I was attempting to generate an instance to reproduce the issue with the help of Eder. However, we've discovered that this problem is more challenging to address during the GPL routability. It does not work to just call global_placement with cells that fall into the issue, as the cells are repositioned before reaching the error stage in GPL routability.

Eder recommended that we proceed with merging the code for the time being and, in the future, provide a unit test for the case where we already know the issue occurs with ihp/Ibex. This could potentially be achieved by incorporating ihp into the test and utilizing DeltaDebugger to isolate the problem.

**Comment by precisionmoon** [comment]
Sorry for the late reply, but this PR fixed my fatal during remap.  Thanks.
